### PR TITLE
Improvement Spending By Category Recipient performance

### PR DESCRIPTION
**Pull request by tony-sappe**
**High level description:**
Main actions performed in this PR:

1. Incorporated new `duns` table to normalize Recipient Names by DUNS
1. Created a new materialized view for recipients: `summary_transaction_recipient_view`
1. Rolled up Recipient Location by `transaction_fpds` or `transaction_fabs` instead of using the `reference_location` table
1. Dropped unused indexes from the DB

**Technical details:**

- DUNS table is a direct transfer from the broker performed in PR #1302 
- The location data changes were from modifying how the joining is performed. By reducing the joins the intent was to reduce the temporary memory and storage requirements while creating materialized views. Preliminary testing showed that data quality might also be improved.
- A number of indexes had been noticed for months as not being used. In an effort to improve overall performance (ETL, query planner, etc) some of them were dropped to reduce overall bloat and storage needs. Not all indexes were removed, rather only a few from the tables and some materialized views. The SQL used to determine unused indexes is below. Note, it **must** be run on production.
```
SELECT s.schemaname,
       s.relname AS tablename,
       s.indexrelname AS indexname,
        pg_size_pretty(pg_relation_size(s.indexrelid)) AS index_size
FROM pg_catalog.pg_stat_user_indexes s
   JOIN pg_catalog.pg_index i ON s.indexrelid = i.indexrelid
WHERE s.idx_scan = 0      -- has never been scanned
  AND 0 <>ALL (i.indkey)  -- no index column is an expression
  AND NOT EXISTS          -- does not enforce a constraint
         (SELECT 1 FROM pg_catalog.pg_constraint c
          WHERE c.conindid = s.indexrelid)
ORDER BY pg_relation_size(s.indexrelid) DESC;
```

**Link to JIRA Ticket:**
[DEV-1169](https://federal-spending-transparency.atlassian.net/browse/DEV-1169)

**The following are ALL required for the PR to be merged:**
- [x] Backend review completed
- [x] Unit & integration tests updated with relevant test cases
- [x] Matview impact assessment completed
- [x] Frontend impact assessment completed
- [x] Data validation completed
- [x] API Performance evaluation completed and present

### CODE REVIEW

File: usaspending_api/search/v2/views/spending_by_category.py [hunk 1]
```diff
@@ -216,10 +218,25 @@ def recipient(self) -> list:
...
+                lookup = RecipientLookup.objects \
+                    .filter(recipient_hash=row['recipient_hash']) \
+                    .values('legal_business_name', 'duns').first()
+                if lookup:
+                    row['recipient_name'] = lookup.get('legal_business_name', None)
+                    row['recipient_unique_id'] = lookup.get('duns', None)
+                else:
+                    row['recipient_name'] = 'DUNS not Provided'
+                    row['recipient_unique_id'] = None
+                row['recipient_id'] = None
```
[the remark this answers is not in the document]
Comment (tony-sappe) [reply usaspending_api/search/v2/views/spending_by_category.py]: Will change.

Discovered there are many "categorized non-DUNS" recipients:
* MULTIPLE RECIPIENTS
* REDACTED DUE TO PII
* MULTIPLE FOREIGN RECIPIENTS
* PRIVATE INDIVIDUAL
* INDIVIDUAL RECIPIENT

Currently only `MULTIPLE RECIPIENTS` is handled well in this PR. Will add more entires for these items.

[the remark this answers is not in the document]
Comment (dpb-bah) [reply usaspending_api/search/v2/views/spending_by_category.py]: Great catch! I've never encountered those.

**Comment by tony-sappe** [comment]
### Performance Benchmarks
DUNS, FY2014-2018 POP
Staging: 12s
New:  - 3.2s

{"category":"recipient_duns","filters":{"time_period":[{"start_date":"2015-10-01","end_date":"2017-09-30"}]},"limit":10,"page":1,"auditTrail":"Recipient Rank Visualization","subawards":false}
Staging: 40s
New: 6.8s

{"category":"recipient_duns","filters":{"keyword":"lockheed","time_period":[{"start_date":"2014-10-01","end_date":"2017-09-30"}]},"limit":10,"page":1,"auditTrail":"Recipient Rank Visualization","subawards":false}
Staging: 12s
New: 2.9s

{"category":"recipient_duns","filters":{"time_period":[{"start_date":"2007-10-01","end_date":"2018-09-30"}]},"limit":10,"page":1,"auditTrail":"Recipient Rank Visualization","subawards":false}
Staging: TIMEOUT
New: 12.1s

{"category":"recipient_duns","filters":{"time_period":[{"start_date":"2009-10-01","end_date":"2018-09-30"}]},"limit":10,"page":1,"auditTrail":"Recipient Rank Visualization","subawards":false}
Staging: 43s
New: 10.7s


{"category":"recipient_duns","filters":{"time_period":[{"start_date":"2017-10-01","end_date":"2018-09-30"},{"start_date":"2016-10-01","end_date":"2017-09-30"},{"start_date":"2015-10-01","end_date":"2016-09-30"}]},"limit":10,"page":1,"auditTrail":"Recipient Rank Visualization","subawards":false}
Staging: 36.4s
New: 7.1s


{"category":"recipient_duns","filters":{"time_period":[{"start_date":"2007-10-01","end_date":"2018-09-30"}],"place_of_performance_locations":[{"country":"USA","state":"PR"}]},"limit":10,"page":1,"auditTrail":"Recipient Rank Visualization","subawards":false}
Staging: Timeout
New: 2.7s

**Comment by tony-sappe** [comment]
### Materialized View Impact
There are two new materialized views; one is a lookup table for DUNS and the other is for improving the API responses for Spending By Category|Recipient.

In order to offset the additional time and space to create the matviews a number of unused indexes were dropped. This has the additional benefit of improving the ETL process since there is less overhead for insert/update/deletes to the models.
Another change was to reduce the number of JOINs for the matview creates.

**Pull request merged**